src/NonfungiblePositionManager.sol

```diff
@@ -0,0 +1,226 @@
+    /// @param sender is the msg.sender encoded by the `modifyLiquidities` function before the `unlockCallback`.
+    /// @return int256 the balance of the currency being settled by this call
+    function _close(bytes memory params, address sender) internal returns (bytes memory) {
+        (Currency currency) = abi.decode(params, (Currency));
+        // this address has applied all deltas on behalf of the user/owner
+        // it is safe to close this entire delta because of slippage checks throughout the batched calls.
+        int256 currencyDelta = manager.currencyDelta(address(this), currency);
+
+        // the sender is the payer or receiver
+        if (currencyDelta < 0) {
+            currency.settle(manager, sender, uint256(-int256(currencyDelta)), false);
+        } else {
+            currency.take(manager, sender, uint256(int256(currencyDelta)), false);
+        }
+
+        return abi.encode(currencyDelta);
+    }
+
+    function burn(uint256 tokenId, address sender) internal {
+        _requireApprovedOrOwner(tokenId, sender);
+        // We do not need to enforce the pool manager to be unlocked bc this function is purely clearing storage for the minted tokenId.
+
+        // Checks that the full position's liquidity has been removed and all tokens have been collected from tokensOwed.
+        _validateBurn(tokenId);
+
+        delete tokenRange[tokenId];
+        // Burn the token.
+        _burn(tokenId);
+    }
+
+    function _modifyLiquidity(LiquidityRange memory range, int256 liquidityChange, bytes32 salt, bytes memory hookData)
+        internal
+        returns (BalanceDelta liquidityDelta, BalanceDelta totalFeesAccrued)
+    {
+        (liquidityDelta, totalFeesAccrued) = manager.modifyLiquidity(
+            range.poolKey,
+            IPoolManager.ModifyLiquidityParams({
+                tickLower: range.tickLower,
+                tickUpper: range.tickUpper,
+                liquidityDelta: liquidityChange,
+                salt: salt
+            }),
+            hookData
+        );
+    }
+
+    function _validateBurn(uint256 tokenId) internal {
+        bytes32 positionId = getPositionIdFromTokenId(tokenId);
+        uint128 liquidity = manager.getPositionLiquidity(tokenRange[tokenId].poolKey.toId(), positionId);
+        if (liquidity > 0) revert PositionMustBeEmpty();
+    }
+
+    // TODO: Move this to a posm state-view library.
+    function getPositionIdFromTokenId(uint256 tokenId) public view returns (bytes32 positionId) {
+        LiquidityRange memory range = tokenRange[tokenId];
+        bytes32 salt = bytes32(tokenId);
+        int24 tickLower = range.tickLower;
+        int24 tickUpper = range.tickUpper;
+        address owner = address(this);
+
+        // positionId = keccak256(abi.encodePacked(owner, tickLower, tickUpper, salt))
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(0x26, salt) // [0x26, 0x46)
```

Comment:
we now have this logic 3 times in core/periphery we should definitely have it in a helper

Reply:
also this assembly is not memory safe - so you either need to remove the label, or use the free memory pointer.
see the memory safe implementation [here](https://github.com/Uniswap/v4-core/blob/7a72031574fc4548ca8fce197114cf87d5a2c037/src/libraries/Position.sol#L42-L54)

Reply:
There is an outstanding task to move this into a view only library. TODO in the code. For now I removed the memory-safe comment

Reply:
https://github.com/Uniswap/v4-periphery/issues/171